**Project Name**: Fake News Detection using NLP

**Done by**: Y. Divya Reddy

## Problem Statement:

Fake news is a big problem today, making it hard to know what's real. This project aims to build a tool that uses smart computer programs (NLP) to tell if news articles are true or false, helping us fight misinformation.

Install Libraries

In [1]:
!pip install nltk scikit-learn pandas

Import Libraries

In [2]:
import pandas as pd
import numpy as np
import nltk
import re

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Load Dataset

In [4]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

fake = pd.read_csv('/content/drive/MyDrive/Fake.csv')
true = pd.read_csv('/content/drive/MyDrive/True.csv')

Mounted at /content/drive


Display Dataset

In [5]:
fake.head()

,title,text,subject,date
0,Donald Trump Sends Out Embarrassing New Year’...,Donald Trump just couldn t wish all Americans ...,News,"December 31, 2017"
1,Drunk Bragging Trump Staffer Started Russian ...,House Intelligence Committee Chairman Devin Nu...,News,"December 31, 2017"
2,Sheriff David Clarke Becomes An Internet Joke...,"On Friday, it was revealed that former Milwauk...",News,"December 30, 2017"
3,Trump Is So Obsessed He Even Has Obama’s Name...,"On Christmas day, Donald Trump announced that ...",News,"December 29, 2017"
4,Pope Francis Just Called Out Donald Trump Dur...,Pope Francis used his annual Christmas Day mes...,News,"December 25, 2017"


In [6]:
true.head()

,title,text,subject,date
0,"As U.S. budget fight looms, Republicans flip t...",WASHINGTON (Reuters) - The head of a conservat...,politicsNews,"December 31, 2017"
1,U.S. military to accept transgender recruits o...,WASHINGTON (Reuters) - Transgender people will...,politicsNews,"December 29, 2017"
2,Senior U.S. Republican senator: 'Let Mr. Muell...,WASHINGTON (Reuters) - The special counsel inv...,politicsNews,"December 31, 2017"
3,FBI Russia probe helped by Australian diplomat...,WASHINGTON (Reuters) - Trump campaign adviser ...,politicsNews,"December 30, 2017"
4,Trump wants Postal Service to charge 'much mor...,SEATTLE/WASHINGTON (Reuters) - President Donal...,politicsNews,"December 29, 2017"


Rows and column count

In [7]:
print(fake.shape)
print(true.shape)

(23481, 4)
(21417, 4)


Data cleaning

Missing values

In [8]:
fake.isnull().sum()

,0
title,0
text,0
subject,0
date,0


In [9]:
true.isnull().sum()

,0
title,0
text,0
subject,0
date,0


Labeling

In [10]:
fake['label'] = 0
true['label'] = 1

Merge both datasets

In [11]:
data = pd.concat([fake, true], axis=0)

Shuffle dataset

In [12]:
#to avoid bias
data = data.sample(frac=1)

Reset index

In [13]:
data.reset_index(inplace=True)

Feature selection

In [14]:
data = data[['text', 'label']]

In [15]:
data.head()

,text,label
0,Donald Trump s relationship with the press is ...,0
1,The case of a New York Conservative Party lead...,0
2,President Donald Trump didn t take long to put...,0
3,Donald Trump has been obsessed with the fact t...,0
4,VIENNA (Reuters) - Austrian Finance Minister H...,1


Create Text Cleaning Function

In [16]:
def clean_text(text):

    # Convert to lowercase
    text = text.lower()

    # Remove URLs
    text = re.sub(r'http\S+', '', text)

    # Remove special characters and numbers
    text = re.sub(r'[^a-zA-Z]', ' ', text)

    # Remove extra spaces
    text = text.split()
    text = ' '.join(text)

    return text

Apply cleaning function

In [17]:
data['text'] = data['text'].apply(clean_text)

In [18]:
data.head()

,text,label
0,donald trump s relationship with the press is ...,0
1,the case of a new york conservative party lead...,0
2,president donald trump didn t take long to put...,0
3,donald trump has been obsessed with the fact t...,0
4,vienna reuters austrian finance minister hans ...,1


Remove duplicate rows

In [19]:
data.drop_duplicates(inplace=True)

checking class distribution

In [20]:
data['label'].value_counts()

,count
label,
1,20922
0,17393


Split data

In [21]:
X = data['text']
y = data['label']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

TF-IDF Vectorization

In [22]:
vectorizer = TfidfVectorizer(max_features=5000)

X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

Train model

In [23]:
model = LogisticRegression()

model.fit(X_train, y_train)

LogisticRegression()

Accuracy

In [24]:
pred = model.predict(X_test)

accuracy = accuracy_score(y_test, pred)

print("Accuracy:", accuracy)

Accuracy: 0.9835573535168994


Save model for deployment

In [25]:
import pickle

pickle.dump(model, open('model.pkl', 'wb'))
pickle.dump(vectorizer, open('vectorizer.pkl', 'wb'))

Download files

In [26]:
from google.colab import files

# Download model file
files.download('model.pkl')

# Download vectorizer file
files.download('vectorizer.pkl')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Project Summary

This project developed a Fake News Detection system using NLP. We preprocessed news articles, used TF-IDF for text vectorization, and trained a Logistic Regression model. The model achieved 98.35% accuracy in classifying articles as fake or true.

## Conclusion

We successfully built an accurate fake news detection model with over 98% accuracy using Logistic Regression and TF-IDF. This model is a valuable tool against misinformation. Future work could involve advanced NLP models and larger datasets.